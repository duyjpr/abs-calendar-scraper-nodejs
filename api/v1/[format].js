import { parse } from "node-html-parser";
import fetch from "node-fetch";
import ical from "ical-generator";

const ABS_ROOT = "https://www.abs.gov.au/";
const FUTURE_RELEASES = "https://www.abs.gov.au/release-calendar/future-releases";
const PRODID = "//github:duyjpr//abs-calendar-scraper-nodejs//EN";

async function loadPage(url) {
  const doc = await fetch(url);
  const contents = await doc.text();
  return parse(contents);
}

function absLink(href) {
  if(!href.startsWith(ABS_ROOT)) {
    href = ABS_ROOT + href;
  }
  return href;
}

async function getMonthlyUrls(url) {
  const pg = await loadPage(url);
  const linkNodes = pg.querySelector(".date-pager").querySelectorAll("a[href]");
  return linkNodes.map(a => absLink(a.attributes["href"]));
}

function parentUrl(url) {
  return url.substr(0, url.lastIndexOf("/"));
}

async function getReleases(url) {
  const doc = await loadPage(url);
  const entries = doc.querySelectorAll(".views-row");
  const times = entries.map(e => {
    return e.querySelector(".event-date time").attributes["datetime"];
  });
  const titles = entries.map(e => {
    return e.querySelector(".event-name").textContent.trim();
  });
  const descriptions = entries.map(e => {
    return e.querySelector(".event-description").textContent.trim();
  });
  const referencePeriods = entries.map(e => {
    return e.querySelector(".reference-period-value").textContent.trim();
  });
  const latestUrls = entries.map(e => {
    const a = e.querySelector(".rs-product-link-latest a[href]");
    return !a
      ? null
      : absLink(parentUrl(a.attributes["href"].trim()) + "/latest-release");
  });
  const topics = latestUrls.map(u => {
    if(!u) return null;
    const parts = u.split("/");
    // Terminology comes from ABS website.
    // Format: abs.gov.au/statistics/theme/parent-topic/topic/latest-release 
    // https://www.abs.gov.au/welcome-new-abs-website
    return {
      theme: parts.at(-4),
      parentTopic: parts.at(-3),
      topic: parts.at(-2)
    };
  });

  const entriesOnPage = times.map((t, i) => ({
    time: new Date(times[i]),
    title: titles[i],
    description: descriptions[i],
    referencePeriod: referencePeriods[i],
    latestUrl: latestUrls[i],
    ...topics[i]
  }));
  return entriesOnPage;
}

function matchesFilters(release, filters) {
  for(let [name, values] of filters) {
    const checkValue = release[name];
    // Edge case: single filter value
    if(checkValue === values) return true;

    for(let value of values) {
      if(checkValue === value) return true;
    }
  }
  return false;
}

function toIcal(entries) {
  const calendar = ical({
    prodId: PRODID
  });
  for(let e of entries) {
    calendar.createEvent({
      start: e.time,
      end: e.time,
      summary: e.title + ", " + e.referencePeriod,
      description: [
        e.description,
        "",
        !e.theme ? "" : `${e.theme} / ${e.parentTopic} / ${e.topic}`,
        e.latestUrl
      ].join("\n")
    });
  }
  return calendar;
}

export default async function handler(request, response) {
  const monthlyUrls = await getMonthlyUrls(FUTURE_RELEASES);
  const monthlyEntries = await Promise.all(monthlyUrls.map(getReleases));
  let releases = monthlyEntries.flat();

  const {format = "json", ...filters} = request.query;

  // [field, value | [value1...valueN]]
  const filterEntries = Object.entries(filters);
  if(filterEntries.length > 0) {
    releases = releases.filter(r => matchesFilters(r, filterEntries));
  }

  if(format === "icalendar") {
    return toIcal(releases).serve(response);
  }
  response.status(200).json(releases);
}
