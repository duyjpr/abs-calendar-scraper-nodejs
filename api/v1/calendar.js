// Choose Vercel edge runtime
export const config = {
  runtime: "edge",
};

import ICAL from "ical.js";
import { parse } from "node-html-parser";

const ABS_ROOT = "https://www.abs.gov.au/";
const ABS_CALENDAR = ABS_ROOT + "release-calendar/future-releases-calendar";
const PRODID = "//github:duyjpr//abs-calendar-scraper-nodejs//EN";
const CALENDAR_NAME = "ABS releases";

export async function GET(request) {
  // Copy out of URLSearchParams into a plain object that works with
  // destructuring assignment. JS doesn't have a standard way of handling
  // repeated searchParams so we have to do things manually.
  const query = {};
  for(const [k, v] of new URL(request.url).searchParams) {
    const testValue = query[k];
    if(testValue === undefined) {
      query[k] = v;
    } else if(Array.isArray(testValue)) {
      testValue.push(v);
    } else {
      query[k] = [query[k], v];
    }
  }

  const {
    format = "icalendar",
    allday = "Australia/Melbourne",
    ...filterInputs
  } = query;
  const filters = Object.entries(filterInputs); // {name, value | [values]}

  let releases = await scrapeAbsCalendar();
  if(filters.length > 0) {
    releases = releases.filter(rls => {
      for(const [name, values] of filters) {
        const target = rls[name];
        const found = (values === target) ||
          (values.find && values.find(x => x === target));
        // Early exit if match found
        if(found) return true;
      }
      return false;
    });
  }

  if(format === "icalendar") {
    return new Response(
      releasesToIcal(releases, allday === "false" ? undefined : allday),
      {
        headers: {
          "Content-Type": "text/calendar;charset=utf-8",
          "Content-Disposition": 'attachment; filename="abs-calendar.ics"',
        },
      },
    );
  }

  return Response.json(releases);
}

async function parseFromUrl(url) {
  const doc = await fetch(url);
  const raw = await doc.text();
  return parse(raw);
}

async function scrapeAbsCalendar() {
  const monthlyUrls = (await parseFromUrl(ABS_CALENDAR))
    .querySelector(".date-pager")
    .querySelectorAll("a[href]")
    .map(a => new URL(a.attributes.href, ABS_CALENDAR).href);
  const monthlyReleases = await Promise.all(
    monthlyUrls.map(scrapeReleasesOnPage)
  );
  return monthlyReleases.flat();
}

function text(x) {
  return x.textContent.trim();
}

async function scrapeReleasesOnPage(u) {
  const page = await parseFromUrl(u);
  const entries = page.querySelectorAll(".calendar");
  return entries.map(e => ({
    time: e.querySelector("time").attributes.datetime,
    title: text(e.querySelector(".event-name")),
    referencePeriod: text(e.querySelector(".reference-period-value")),

    ...scrapeLatestUrlAndTopics(
      e.querySelector(".rs-product-link-latest a[href]")?.attributes.href
    )
  }));
}

function scrapeLatestUrlAndTopics(href) {
  if(href) {
    const latestUrl = new URL(
      href.replace(new RegExp("/[^/]+$", "u"), "/latest-release"),
      ABS_CALENDAR
    ).href;
    const parts = latestUrl.split("/");
    // Terminology comes from ABS website.
    // Format: abs.gov.au/statistics/theme/parent-topic/topic/latest-release 
    // https://www.abs.gov.au/welcome-new-abs-website
    const topics = {
      theme: parts.at(-4),
      parentTopic: parts.at(-3),
      topic: parts.at(-2)
    };

    return {latestUrl, ...topics};
  }
  return null;
}

function releasesToIcal(releases, allDay) {
  const now = ICAL.Time.fromJSDate(new Date(), true);
  const calendar = new ICAL.Component("vcalendar");
  calendar.addPropertyWithValue("version", "2.0");
  calendar.addPropertyWithValue("prodid", PRODID);
  calendar.addPropertyWithValue("x-wr-calname", CALENDAR_NAME);
  // Proposed standardisation, not widely supported
  calendar.addPropertyWithValue("name", CALENDAR_NAME);
  
  for(const rls of releases) {
    const vevent = new ICAL.Component("vevent", calendar);
    const event = new ICAL.Event(vevent);
    event.uid = [
      rls.title, rls.referencePeriod, rls.time
    ].join(" ").replace(/[^A-Za-z0-9]+/gu, "-");
    vevent.addPropertyWithValue("dtstamp", now);
    
    if(allDay) {
      // Convert day to local timezone in case ABS ever releases something
      // before 11am
      const parts = new Date(rls.time)
        .toLocaleDateString("en-AU", {
          timeZone: allDay
        }) // dd/mm/yyyy
        .split("/");

      const time = ICAL.Time.fromData({
        day: Number(parts[0]),
        month: Number(parts[1]),
        year: Number(parts[2]),
        isDate: true,
      });
      event.startDate = time;
    } else {
      const time = ICAL.Time.fromString(rls.time);
      event.startDate = time;
      event.endDate = time;
    }

    event.summary = rls.title + " for " + rls.referencePeriod;
    if(rls.latestUrl) event.description = rls.latestUrl;
    calendar.addSubcomponent(vevent);
  }

  return calendar.toString();
}