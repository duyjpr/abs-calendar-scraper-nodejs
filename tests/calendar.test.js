import { beforeAll, expect, jest, test } from "@jest/globals";
import { GET } from "../api/v1/calendar";
import { load as loadYaml} from "js-yaml";
import { readFile } from "node:fs/promises";

const mockedFetches = loadYaml(
  await readFile("tests/calendar-mocked-fetches.yaml")
);

const mockedFetch = jest.fn(url =>
  Promise.resolve({
    text: () => Promise.resolve(mockedFetches[url]),
  })
);

beforeAll(() => {
  globalThis.fetch = mockedFetch;
});

// Factor out common checking pattern
async function checkOutput(url) {
  const response = await GET(new Request(url));
  let generated = await response.text();
  const mime = response.headers.get("Content-Type")
    .match(/^[^;]*/u)[0];
  if(mime === "text/calendar") {
    generated = generated.replace(/^DTSTAMP:.+$/mgu,"DTSTAMP:20000101T000001Z");
  }

  expect(generated).toMatchSnapshot();
}

test(".ics for all events", async () => await checkOutput(
  "https://test.case",
));

test("filter", async () => await checkOutput(
  "https://test.case?format=icalendar&parentTopic=decade",
));

test("filters", async () => await checkOutput(
  "https://test.case?format=icalendar&topic=start-decade&topic=end-decade&topic=mid-decade",
));

test("filters on multiple fields", async () => await checkOutput(
  "https://test.case?format=icalendar&topic=start-decade&referencePeriod=Dec%202025",
));

test("json output", async () => await checkOutput(
  "https://test.case?format=json",
));

test("exact time", async () => await checkOutput(
  "https://test.case?format=icalendar&allday=false",
));

test("timezone", async () => await checkOutput(
  "https://test.case?format=icalendar&topic=timezone",
));

test("another timezone", async () => await checkOutput(
  "https://test.case?format=icalendar&allday=Australia/Perth&topic=timezone",
));
