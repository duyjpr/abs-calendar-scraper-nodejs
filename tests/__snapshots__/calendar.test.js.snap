// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.ics for all events 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID://github:duyjpr//abs-calendar-scraper-nodejs//EN
X-WR-CALNAME:ABS releases
NAME:ABS releases
BEGIN:VEVENT
UID:Start-decade-normal-event-2000s-2000-01-01T00-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20000101
SUMMARY:Start decade (normal event) for 2000s
DESCRIPTION:https://test.case/dates/decade/start-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:Mid-decade-for-filters-2000s-2004-12-31T23-59-59Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20050101
SUMMARY:Mid decade (for filters) for 2000s
DESCRIPTION:https://test.case/dates/decade/mid-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:End-decade-for-filters-2000s-2009-12-31T23-59-59Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20100101
SUMMARY:End decade (for filters) for 2000s
DESCRIPTION:https://test.case/dates/decade/end-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:Start-decade-2nd-release-2010s-2010-01-01T00-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20100101
SUMMARY:Start decade (2nd release) for 2010s
DESCRIPTION:https://test.case/dates/decade/start-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:New-product-first-release-Dec-2025-2026-01-14T01-30-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20260114
SUMMARY:New product (first release) for Dec 2025
END:VEVENT
BEGIN:VEVENT
UID:Timezone-test-4-July-1am-in-Melbourne-2025-07-03T15-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20250704
SUMMARY:Timezone test for 4 July @ 1am in Melbourne
DESCRIPTION:https://test.case/dates/edge/timezone/latest-release
END:VEVENT
END:VCALENDAR"
`;

exports[`another timezone 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID://github:duyjpr//abs-calendar-scraper-nodejs//EN
X-WR-CALNAME:ABS releases
NAME:ABS releases
BEGIN:VEVENT
UID:Timezone-test-4-July-1am-in-Melbourne-2025-07-03T15-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20250703
SUMMARY:Timezone test for 4 July @ 1am in Melbourne
DESCRIPTION:https://test.case/dates/edge/timezone/latest-release
END:VEVENT
END:VCALENDAR"
`;

exports[`exact time 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID://github:duyjpr//abs-calendar-scraper-nodejs//EN
X-WR-CALNAME:ABS releases
NAME:ABS releases
BEGIN:VEVENT
UID:Start-decade-normal-event-2000s-2000-01-01T00-00-00Z
DTSTAMP:20000101T000001Z
DTSTART:20000101T000000Z
DTEND:20000101T000000Z
SUMMARY:Start decade (normal event) for 2000s
DESCRIPTION:https://test.case/dates/decade/start-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:Mid-decade-for-filters-2000s-2004-12-31T23-59-59Z
DTSTAMP:20000101T000001Z
DTSTART:20041231T235959Z
DTEND:20041231T235959Z
SUMMARY:Mid decade (for filters) for 2000s
DESCRIPTION:https://test.case/dates/decade/mid-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:End-decade-for-filters-2000s-2009-12-31T23-59-59Z
DTSTAMP:20000101T000001Z
DTSTART:20091231T235959Z
DTEND:20091231T235959Z
SUMMARY:End decade (for filters) for 2000s
DESCRIPTION:https://test.case/dates/decade/end-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:Start-decade-2nd-release-2010s-2010-01-01T00-00-00Z
DTSTAMP:20000101T000001Z
DTSTART:20100101T000000Z
DTEND:20100101T000000Z
SUMMARY:Start decade (2nd release) for 2010s
DESCRIPTION:https://test.case/dates/decade/start-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:New-product-first-release-Dec-2025-2026-01-14T01-30-00Z
DTSTAMP:20000101T000001Z
DTSTART:20260114T013000Z
DTEND:20260114T013000Z
SUMMARY:New product (first release) for Dec 2025
END:VEVENT
BEGIN:VEVENT
UID:Timezone-test-4-July-1am-in-Melbourne-2025-07-03T15-00-00Z
DTSTAMP:20000101T000001Z
DTSTART:20250703T150000Z
DTEND:20250703T150000Z
SUMMARY:Timezone test for 4 July @ 1am in Melbourne
DESCRIPTION:https://test.case/dates/edge/timezone/latest-release
END:VEVENT
END:VCALENDAR"
`;

exports[`filter 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID://github:duyjpr//abs-calendar-scraper-nodejs//EN
X-WR-CALNAME:ABS releases
NAME:ABS releases
BEGIN:VEVENT
UID:Start-decade-normal-event-2000s-2000-01-01T00-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20000101
SUMMARY:Start decade (normal event) for 2000s
DESCRIPTION:https://test.case/dates/decade/start-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:Mid-decade-for-filters-2000s-2004-12-31T23-59-59Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20050101
SUMMARY:Mid decade (for filters) for 2000s
DESCRIPTION:https://test.case/dates/decade/mid-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:End-decade-for-filters-2000s-2009-12-31T23-59-59Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20100101
SUMMARY:End decade (for filters) for 2000s
DESCRIPTION:https://test.case/dates/decade/end-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:Start-decade-2nd-release-2010s-2010-01-01T00-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20100101
SUMMARY:Start decade (2nd release) for 2010s
DESCRIPTION:https://test.case/dates/decade/start-decade/latest-release
END:VEVENT
END:VCALENDAR"
`;

exports[`filters 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID://github:duyjpr//abs-calendar-scraper-nodejs//EN
X-WR-CALNAME:ABS releases
NAME:ABS releases
BEGIN:VEVENT
UID:Start-decade-normal-event-2000s-2000-01-01T00-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20000101
SUMMARY:Start decade (normal event) for 2000s
DESCRIPTION:https://test.case/dates/decade/start-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:Mid-decade-for-filters-2000s-2004-12-31T23-59-59Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20050101
SUMMARY:Mid decade (for filters) for 2000s
DESCRIPTION:https://test.case/dates/decade/mid-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:End-decade-for-filters-2000s-2009-12-31T23-59-59Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20100101
SUMMARY:End decade (for filters) for 2000s
DESCRIPTION:https://test.case/dates/decade/end-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:Start-decade-2nd-release-2010s-2010-01-01T00-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20100101
SUMMARY:Start decade (2nd release) for 2010s
DESCRIPTION:https://test.case/dates/decade/start-decade/latest-release
END:VEVENT
END:VCALENDAR"
`;

exports[`filters on multiple fields 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID://github:duyjpr//abs-calendar-scraper-nodejs//EN
X-WR-CALNAME:ABS releases
NAME:ABS releases
BEGIN:VEVENT
UID:Start-decade-normal-event-2000s-2000-01-01T00-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20000101
SUMMARY:Start decade (normal event) for 2000s
DESCRIPTION:https://test.case/dates/decade/start-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:Start-decade-2nd-release-2010s-2010-01-01T00-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20100101
SUMMARY:Start decade (2nd release) for 2010s
DESCRIPTION:https://test.case/dates/decade/start-decade/latest-release
END:VEVENT
BEGIN:VEVENT
UID:New-product-first-release-Dec-2025-2026-01-14T01-30-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20260114
SUMMARY:New product (first release) for Dec 2025
END:VEVENT
END:VCALENDAR"
`;

exports[`json output 1`] = `"[{"time":"2000-01-01T00:00:00Z","title":"Start decade (normal event)","referencePeriod":"2000s","latestUrl":"https://test.case/dates/decade/start-decade/latest-release","theme":"dates","parentTopic":"decade","topic":"start-decade"},{"time":"2004-12-31T23:59:59Z","title":"Mid decade (for filters)","referencePeriod":"2000s","latestUrl":"https://test.case/dates/decade/mid-decade/latest-release","theme":"dates","parentTopic":"decade","topic":"mid-decade"},{"time":"2009-12-31T23:59:59Z","title":"End decade (for filters)","referencePeriod":"2000s","latestUrl":"https://test.case/dates/decade/end-decade/latest-release","theme":"dates","parentTopic":"decade","topic":"end-decade"},{"time":"2010-01-01T00:00:00Z","title":"Start decade (2nd release)","referencePeriod":"2010s","latestUrl":"https://test.case/dates/decade/start-decade/latest-release","theme":"dates","parentTopic":"decade","topic":"start-decade"},{"time":"2026-01-14T01:30:00Z","title":"New product (first release)","referencePeriod":"Dec 2025"},{"time":"2025-07-03T15:00:00Z","title":"Timezone test","referencePeriod":"4 July @ 1am in Melbourne","latestUrl":"https://test.case/dates/edge/timezone/latest-release","theme":"dates","parentTopic":"edge","topic":"timezone"}]"`;

exports[`timezone 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID://github:duyjpr//abs-calendar-scraper-nodejs//EN
X-WR-CALNAME:ABS releases
NAME:ABS releases
BEGIN:VEVENT
UID:Timezone-test-4-July-1am-in-Melbourne-2025-07-03T15-00-00Z
DTSTAMP:20000101T000001Z
DTSTART;VALUE=DATE:20250704
SUMMARY:Timezone test for 4 July @ 1am in Melbourne
DESCRIPTION:https://test.case/dates/edge/timezone/latest-release
END:VEVENT
END:VCALENDAR"
`;
